linbox/matrix/polynomial-matrix.h

```diff
@@ -45,242 +44,275 @@ uint64_t max_memory=0, cur_memory=0;
-	
+    
+    template<class MatPoly>
+	class SubPolynomialMatrix;
+    
+    
 	// Class for Polynomial Matrix stored as a Matrix of Polynomials
 	template<class _Field>
-	class PolynomialMatrix<PMType::polfirst,PMStorage::plain,_Field> {
-	public:
-		typedef _Field Field;
-		typedef typename Field::Element   Element;
-		typedef BlasMatrix<Field>          Matrix;
-		//typedef typename std::vector<Element>::iterator  Iterator;
-		//typedef typename std::vector<Element>::const_iterator  ConstIterator;
+	class PolynomialMatrix<_Field, PMType::polfirst> {
+    public:
+        typedef _Field                         Field;
+		typedef typename Field::Element      Element;
 		typedef std::vector<Element,AlignedAllocator<Element, Alignment::DEFAULT>> VECT;
-		typedef typename VECT::iterator  Iterator;
-		typedef typename VECT::const_iterator  ConstIterator;
-		//typedef vector<Element>        Polynomial;
-		typedef Subvector<Iterator,ConstIterator>   Polynomial;
-		typedef PolynomialMatrix<PMType::polfirst,PMStorage::plain,_Field>  Self_t;
-		typedef PolynomialMatrix<PMType::matfirst,PMStorage::plain,_Field> Other_t;
+        typedef BlasVector<Field>                             Data;        
 
-		//PolynomialMatrix() {}
+        typedef BlasMatrix<Field>                                   Matrix;
+        typedef const BlasMatrix<Field>                        constMatrix;
+        typedef BlasSubvector<Data>                             Polynomial;
+        typedef const BlasSubvector<const Data>            constPolynomial;        
 
+		typedef PolynomialMatrix<Field, PMType::polfirst>  Self_t;
+        typedef SubPolynomialMatrix<Self_t>                  view;
+        typedef SubPolynomialMatrix<const Self_t>      const_view;
+        
 		// construct a polynomial matrix in f[x]^(m x n) of degree (s-1)
-		PolynomialMatrix(const Field& f, size_t r, size_t c, size_t s, size_t stor=0) :
-			_store((stor?stor:s)), _repview(r*c),_rep(r*c*_store,f.zero), _row(r), _col(c), _size(s), _fld(&f) {
-			for (size_t i=0;i<_row;i++)
-				for (size_t j=0;j<_col;j++)
-					_repview[i*_col+j]= Polynomial(_rep.begin()+(i*_col+j)*_store,_size);
-			//integer p;
-			//std::cout<<"MatrixP allocating : "<<r*c*s*length(f.characteristic(p))/1000000.<<"Mo"<<std::endl;
-			//std::cout<<"(ALLOC) PolynomialMatrix<polfirst> at "<<this<<" : "<<r<<"x"<<c<<" - size= "<<s<<" ==> "<<MB(realmeminfo())<<" Mo   "<<STR_MEMINFO<<std::endl;
-			ADD_MEM(realmeminfo());
+		PolynomialMatrix(const Field& f, size_t r, size_t c, size_t s)
+            : _rep(f,r*c*s), _row(r), _col(c), _size(s) {}
+
+
+        /*******************************
+         * Data access functionnalities*
+         ******************************/
+        
+        // retrieve the polynomial at entry (i,j) in the matrix
+        Polynomial       operator()(size_t i, size_t j)      {return operator()(i*_col+j);}
```

Comment:
don't we want to share the same API as the dense matrices (`getEntry` and `refEntry`) ? I understand that ref and get should have distinct names in order to avoid confusion with getEntry(i,j)  when using only 2 parameters, but it would be better to fit the dense matrix API in order to use polynomial matrices as any other matrices for some algorithms.

Reply:
see discussion above https://github.com/linbox-team/linbox/pull/285#discussion_r766559416